include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,616 @@
+   * that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation.
+   * @return true if the allocation is found, false otherwise.
+   */
+  bool deallocate(void* p, std::size_t bytes, cudaStream_t stream)
+  {
+    lock_guard lock(mtx_);
+    bool found = do_deallocate(p, bytes);
+    shrink_arena(stream);
+    return found;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`, keeping all free superblocks.
+   *
+   * This is done when deallocating from another arena. Since we don't have access to the CUDA
+   * stream associated with this arena, it's not safe to return superblocks.
+   *
+   * @param p Pointer to be deallocated.
+   * @param bytes The size in bytes of the allocation. This must be equal to the value of `bytes`
+   * that was passed to the `allocate` call that returned `p`.
+   * @return true if the allocation is found, false otherwise.
+   */
+  bool deallocate(void* p, std::size_t bytes)
+  {
+    lock_guard lock(mtx_);
+    return do_deallocate(p, bytes);
+  }
+
+  /**
+   * @brief Does an arena handle blocks of this size?
+   *
+   * @param size The size in bytes to check.
+   * @return true if blocks of this size are handled by an arena.
+   */
+  static bool handles_size(std::size_t size) { return size <= maximum_allocation_size; }
+
+ private:
+  /// The maximum allocation size handled by arenas.
+  static constexpr std::size_t maximum_allocation_size = superblock_size / 2;
+
+  using lock_guard = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Get an available memory block of at least `size` bytes.
+   *
+   * @param size The number of bytes to allocate.
+   * @return block A block of memory of at least `size` bytes.
+   */
+  block get_block(std::size_t size)
+  {
+    // Find the first-fit free block.
+    auto const b = first_fit(free_blocks_, size);
+    if (b.is_valid()) {
+      free_superblocks_.erase(b.pointer);
+      return b;
+    }
+
+    // No existing larger blocks available, so grow the arena and obtain a superblock.
+    auto const superblock = expand_arena();
+    return split_block(free_blocks_, superblock, size);
+  }
+
+  /**
+   * @brief Allocate space from upstream to supply the arena and return a superblock.
+   *
+   * @return block A superblock.
+   */
+  block expand_arena() const
+  {
+    return {global_arena_.allocate(superblock_size), superblock_size, true};
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated.
+   * @param bytes The size in bytes of the allocation. This must be equal to the value of `bytes`
+   * that was passed to the `allocate` call that returned `p`.
+   * @return bool if the allocation is found.
+   */
+  bool do_deallocate(void* p, std::size_t bytes)
+  {
+    auto const b      = free_block(p, bytes);
+    auto const merged = coalesce_block(free_blocks_, b);
+    if (merged.is_superblock()) { free_superblocks_.insert(merged.pointer); }
```

Comment:
```suggestion
    if (merged.is_superblock() ) { 
      if (free_superblock_.is_valid()) global_arena_.deallocate(merged.pointer(), superblock_size);
      else free_superblock_ = merged;
    }
```

And you can delete the `shrink_arena` function.

Reply:
See above.